packages/capabilities/src/provider.js

```diff
@@ -0,0 +1,60 @@
+/**
+ * Provider Capabilities
+ *
+ * These can be imported directly with:
+ * ```js
+ * import * as Provider from '@web3-storage/capabilities/provider'
+ * ```
+ *
+ * @module
+ */
+import { capability, DID, literal, struct } from '@ucanto/validator'
+import { equalWith, fail, equal } from './utils.js'
+import { top } from './top.js'
+
+export { top }
+
+/**
+ * Capability can only be delegated (but not invoked) allowing audience to
+ * derive any `provider/` prefixed capability for the agent identified
+ * by did:key in the `with` field.
+ */
+export const provider = top.derive({
+  to: capability({
+    can: 'provider/*',
+    with: DID,
+    derives: equalWith,
+  }),
+  derives: equalWith,
+})
+
+const base = top.or(provider)
+
+export const StorageProvider = literal(
+  'did:web:web3.storage:providers:w3up-alpha'
+)
+
+export const AccountDID = DID.match({ method: 'mailto' })
+
+/**
+ * Capability can be invoked by an agent to add a provider to a space.
+ */
+export const add = base.derive({
+  to: capability({
+    can: 'provider/add',
+    with: AccountDID,
+    nb: struct({
+      provider: StorageProvider,
+      consumer: DID.match({ method: 'key' }),
+    }),
+    derives: (child, parent) => {
+      return (
+        fail(equalWith(child, parent)) ||
+        fail(equal(child.nb.provider, parent.nb.provider, 'provider')) ||
+        fail(equal(child.nb.consumer, parent.nb.consumer, 'consumer')) ||
+        true
+      )
+    },
+  }),
+  derives: equalWith,
+})
```

Comment:
We no longer need the `top` and `base` stuff because new ucanto understands `*` patterns natively now.

```suggestion
export const add = capability({
  can: 'provider/add',
  with: AccountDID,
  nb: struct({
    provider: StorageProvider,
    consumer: DID.match({ method: 'key' }),
  }),
  derives: (child, parent) => {
    return (
      fail(equalWith(child, parent)) ||
      fail(equal(child.nb.provider, parent.nb.provider, 'provider')) ||
      fail(equal(child.nb.consumer, parent.nb.consumer, 'consumer')) ||
      true
    )
  },
})
```

Reply:
prettier!

Reply:
https://github.com/web3-storage/w3protocol/pull/462/commits/a199c11a4915b0858473f31e94072e881b8b737c